packages/typir/src/features/inference.ts

```diff
@@ -7,55 +7,86 @@
 import { Type, isType } from '../graph/type-node.js';
 import { Typir } from '../typir.js';
 import { TypirProblem } from '../utils/utils-type-comparison.js';
-import { assertUnreachable } from '../utils/utils.js';
 
 export interface InferenceProblem {
     domainElement: unknown;
     inferenceCandidate?: Type;
     location: string;
-    rule?: TypeInferenceRule; // for debugging only, since rules have no names (so far)
+    rule?: TypeInferenceRule; // for debugging only, since rules have no names (so far); TODO this does not really work with TypeInferenceRuleWithoutInferringChildren
     subProblems: TypirProblem[]; // might be missing or empty
 }
 export function isInferenceProblem(problem: unknown): problem is InferenceProblem {
     return typeof problem === 'object' && problem !== null && typeof (problem as InferenceProblem).location === 'string' && (problem as InferenceProblem).domainElement !== undefined;
 }
 
+// Type and Value to indicate, that an inference rule is intended for another case, and therefore is unable to infer a type for the current case.
+export type InferenceRuleNotApplicable = 'N/A'; // or 'undefined' instead?
+export const InferenceRuleNotApplicable = 'N/A'; // or 'undefined' instead?
+
+type TypeInferenceResultWithoutInferringChildren =
+    /** the identified type */
+    Type |
+    /** 'N/A' to indicate, that the current inference rule is not applicable for the given domain element at all */
+    InferenceRuleNotApplicable |
+    /** a domain element whose type should be inferred instead */
+    unknown |
+    /** an inference problem */
+    InferenceProblem;
+type TypeInferenceResultWithInferringChildren =
+    /** the usual results, since it might be possible to determine the type of the parent without its children */
+    TypeInferenceResultWithoutInferringChildren |
+    /** the children whos types need to be inferred and taken into account to determine the parent's type */
+    unknown[];
+
 /**
  * Represents a single rule for inference,
  * i.e. only a single type (or no type at all) can be inferred for a given domain element.
+ * There are inference rules which dependent on types of children of the given domain element (e.g. calls of overloaded functions depend on the types of the current arguments)
+ * and there are inference rules without this need.
+ */
+export type TypeInferenceRule = TypeInferenceRuleWithoutInferringChildren | TypeInferenceRuleWithInferringChildren;
+
+/** Usual inference rule which don't depend on children's types. */
+export type TypeInferenceRuleWithoutInferringChildren = (domainElement: unknown, typir: Typir) => TypeInferenceResultWithoutInferringChildren;
+
+/**
+ * Inference rule which requires for the type inference of the given parent to take the types of its children into account.
+ * Therefore, the types of the children need to be inferred first.
  */
-export interface TypeInferenceRule {
+export interface TypeInferenceRuleWithInferringChildren {
```

Comment:
Long names sound very specific. Is it really needed?

Reply:
I want to distinguish `TypeInferenceRuleWithoutInferringChildren` and `TypeInferenceRuleWithInferringChildren`. Maybe we could remove the "Type" prefix everywhere?

Reply:
we keep the current names as discussed